server/src/models/swing.rs

```diff
@@ -110,3 +278,332 @@ fn decode(model: &[u8]) -> (f64, f64) {
         )
     }
 }
+
+/// Calculate `a` and `b` of a linear function that intersects with the data
+/// points (first_timestamp, first_value) and (final_timestamp, final_value).
+fn calculate_a_and_b(
+    first_timestamp: Timestamp,
+    first_value: f64,
+    final_timestamp: Timestamp,
+    final_value: f64,
+) -> (f64, f64) {
+    // It seems that no values can be assigned to first_value and final_value
+    // so a * timestamp + b = INFINITY and a * timestamp + b = NEG_INFINITY.
+    if first_value.is_finite() && final_value.is_finite() {
+        let a = (final_value - first_value) / (final_timestamp - first_timestamp) as f64;
+        let b = first_value - a * first_timestamp as f64;
+        (a, b)
+    } else {
+        debug_assert!(equal_or_nan(first_value, final_value));
```

Comment:
The `debug_assert!` was intended as an additional check to help catch bugs that allow the function to be called with two values that are not finite but also not equal. Since `debug_assert!`s are not present in [release builds](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:3,positionColumn:1,positionLineNumber:3,selectionStartColumn:1,selectionStartLineNumber:2,startColumn:1,startLineNumber:2),source:'pub+fn+square(num:+i32)+-%3E+i32+%7B%0A++++debug_assert!!(num+%3D%3D+1)%3B%0A++++num+*+num%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1620,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'rustc+1.62.0+(Rust,+Editor+%231,+Compiler+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4), I think they provide a good way to document and more importantly check assumptions in the code without adding runtime overhead. However, we can also remove the assert if you think that is better?

Reply:
Since the component is still under development I think it is fine. Once the compression is more stable it might be relevant to remove it but for now it seems fine.